[CELEBORN-1425][HELM] Add helm chart unit tests to ensure manifests are rendered as expected  Edit  Add comment

ChenYi015 opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  - Make sure the PR title start w/ a JIRA ticket, e.g. '[CELEBORN-XXXX] Your PR title ...'.
  - Be sure to keep the PR description updated to reflect all changes.
  - Please write your PR title to summarize what this PR proposes.
  - If possible, provide a concise example to reproduce the issue for a faster review.
-->

### What changes were proposed in this pull request?

- Add helm chart unit tests to ensure manifests are rendered as expected.

### Why are the changes needed?

To achieve better robustness and availability.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

I had added some unit test files in `charts/celeborn/tests`. Detailed information about how to run helm chart unit tests can be found here [helm-unittest/helm-unittest](https://github.com/helm-unittest/helm-unittest). First, you need to install helm unit test plugin:

```shell
helm plugin install https://github.com/helm-unittest/helm-unittest.git
```

Then, run helm chart unitt tests as follows:

```shell
$ helm unittest charts/celeborn  --file "tests/**/*_test.yaml" --strict --debug
load_plugins.go:110: [info] file (/Users/chenyi/Library/helm/plugins/helm-acr/completion.yaml) not provided by plugin. No plugin auto-completion possible

### Chart [ celeborn ] charts/celeborn

 PASS  Test Celeborn configmap  charts/celeborn/tests/configmap_test.yaml
 PASS  Test Celeborn master pod monitor charts/celeborn/tests/master/podmonitor_test.yaml
 PASS  Test Celeborn master priority class      charts/celeborn/tests/master/priorityclass_test.yaml
 PASS  Test Celeborn master service     charts/celeborn/tests/master/service_test.yaml
 PASS  Test Celeborn master statefulset charts/celeborn/tests/master/statefulset_test.yaml
 PASS  Test Celeborn worker pod monitor charts/celeborn/tests/worker/podmonitor_test.yaml
 PASS  Test Celeborn worker priority class      charts/celeborn/tests/worker/priorityclass_test.yaml
 PASS  Test Celeborn worker service     charts/celeborn/tests/worker/service_test.yaml
 PASS  Test Celeborn worker statefulset charts/celeborn/tests/worker/statefulset_test.yaml

Charts:      1 passed, 1 total
Test Suites: 9 passed, 9 total
Tests:       46 passed, 46 total
Snapshot:    0 passed, 0 total
Time:        180.841209ms
```

ChenYi015 commented:
/assign @RexXiong

ChenYi015 commented:
I had did some e2e tests as follows:

1. Create a kind cluster with the following config:

```yaml
kind: Cluster
apiVersion: kind.x-k8s.io/v1alpha4
nodes:
- role: control-plane
- role: worker
- role: worker
- role: worker
- role: worker
- role: worker
```

```shell
$ kind create cluster --name celeborn --config kind-config.yaml
Creating cluster "celeborn" ...
 ✓ Ensuring node image (kindest/node:v1.29.2) 🖼
 ✓ Preparing nodes 📦 📦 📦 📦 📦 📦  
 ✓ Writing configuration 📜 
 ✓ Starting control-plane 🕹️ 
 ✓ Installing CNI 🔌 
 ✓ Installing StorageClass 💾 
 ✓ Joining worker nodes 🚜 
Set kubectl context to "kind-celeborn"
You can now use your cluster with:

kubectl cluster-info --context kind-celeborn

Thanks for using kind! 😊
```

2. Build docker image locally and load into kind cluster

```shell
docker build -f docker/Dockerfile -t docker.io/apache/celeborn:v0.4.0-incubating .
kind load docker-image --name celeborn docker.io/apache/celeborn:v0.4.0-incubating
```

3. Install the helm chart

```shell
$ helm install celeborn charts/celeborn \
    --namespace celeborn \
    --create-namespace \
    --set image.repository=apache/celeborn \
    --set image.tag=v0.4.0-incubating \
    --set image.pullPolicy=Never \
    --debug
install.go:200: [debug] Original chart version: ""
install.go:217: [debug] CHART PATH: /Users/chenyi/Code/github.com/apache/celeborn/charts/celeborn

client.go:134: [debug] creating 1 resource(s)
client.go:134: [debug] creating 5 resource(s)
NAME: celeborn
LAST DEPLOYED: Fri May 10 16:58:44 2024
NAMESPACE: celeborn
STATUS: deployed
REVISION: 1
TEST SUITE: None
USER-SUPPLIED VALUES:
image:
  pullPolicy: Never
  repository: apache/celeborn
  tag: v0.4.0-incubating

COMPUTED VALUES:
affinity:
  master:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: app.kubernetes.io/name
            operator: In
            values:
            - celeborn
          - key: app.kubernetes.io/role
            operator: In
            values:
            - master
        topologyKey: kubernetes.io/hostname
  worker:
    podAntiAffinity:
      requiredDuringSchedulingIgnoredDuringExecution:
      - labelSelector:
          matchExpressions:
          - key: app.kubernetes.io/name
            operator: In
            values:
            - celeborn
          - key: app.kubernetes.io/role
            operator: In
            values:
            - worker
        topologyKey: kubernetes.io/hostname
celeborn:
  celeborn.application.heartbeat.timeout: 120s
  celeborn.master.ha.enabled: true
  celeborn.master.http.port: 9098
  celeborn.metrics.enabled: true
  celeborn.push.stageEnd.timeout: 120s
  celeborn.rpc.dispatcher.numThreads: 4
  celeborn.rpc.io.clientThreads: 64
  celeborn.rpc.io.numConnectionsPerPeer: 2
  celeborn.rpc.io.serverThreads: 64
  celeborn.shuffle.chunk.size: 8m
  celeborn.worker.fetch.io.threads: 32
  celeborn.worker.flusher.buffer.size: 256K
  celeborn.worker.heartbeat.timeout: 120s
  celeborn.worker.http.port: 9096
  celeborn.worker.monitor.disk.enabled: false
  celeborn.worker.push.io.threads: 32
cluster:
  name: cluster
dnsPolicy: ClusterFirst
environments:
  CELEBORN_MASTER_JAVA_OPTS: -XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps
    -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced
  CELEBORN_MASTER_MEMORY: 2g
  CELEBORN_NO_DAEMONIZE: 1
  CELEBORN_WORKER_JAVA_OPTS: -XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps
    -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced
  CELEBORN_WORKER_MEMORY: 2g
  CELEBORN_WORKER_OFFHEAP_MEMORY: 12g
  TZ: Asia/Shanghai
fullnameOverride: ""
hostNetwork: false
image:
  pullPolicy: Never
  pullSecrets: []
  repository: apache/celeborn
  tag: v0.4.0-incubating
nameOverride: ""
nodeSelector: {}
podAnnotations: {}
podMonitor:
  enable: true
  podMetricsEndpoint:
    interval: 5s
    portName: metrics
    scheme: http
priorityClass:
  master:
    create: false
    name: ""
    value: 1000000000
  worker:
    create: false
    name: ""
    value: 999999000
replicaCount:
  master: 3
  worker: 5
resources:
  master: {}
  worker: {}
securityContext:
  fsGroup: 10006
  runAsGroup: 10006
  runAsUser: 10006
service:
  port: 9097
  type: ClusterIP
tolerations: []
volumes:
  master:
  - capacity: 100Gi
    hostPath: /mnt/celeborn_ratis
    mountPath: /mnt/celeborn_ratis
    type: hostPath
  worker:
  - capacity: 100Gi
    diskType: SSD
    hostPath: /mnt/disk1
    mountPath: /mnt/disk1
    type: hostPath
  - capacity: 100Gi
    diskType: SSD
    hostPath: /mnt/disk2
    mountPath: /mnt/disk2
    type: hostPath
  - capacity: 100Gi
    diskType: SSD
    hostPath: /mnt/disk3
    mountPath: /mnt/disk3
    type: hostPath
  - capacity: 100Gi
    diskType: SSD
    hostPath: /mnt/disk4
    mountPath: /mnt/disk4
    type: hostPath

HOOKS:
MANIFEST:
---
# Source: celeborn/templates/configmap.yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: celeborn-conf
  labels:
    helm.sh/chart: celeborn-0.1.0
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/version: "0.4.0"
    app.kubernetes.io/managed-by: Helm
data:
  celeborn-defaults.conf: |-
    celeborn.master.endpoints=celeborn-master-0.celeborn-master-svc.celeborn.svc.cluster.local,celeborn-master-1.celeborn-master-svc.celeborn.svc.cluster.local,celeborn-master-2.celeborn-master-svc.celeborn.svc.cluster.local,
    celeborn.master.ha.node.0.host=celeborn-master-0.celeborn-master-svc.celeborn.svc.cluster.local
    celeborn.master.ha.node.1.host=celeborn-master-1.celeborn-master-svc.celeborn.svc.cluster.local
    celeborn.master.ha.node.2.host=celeborn-master-2.celeborn-master-svc.celeborn.svc.cluster.local
    celeborn.master.ha.ratis.raft.server.storage.dir=/mnt/celeborn_ratis
    celeborn.worker.storage.dirs=/mnt/disk1:disktype=SSD:capacity=100Gi,/mnt/disk2:disktype=SSD:capacity=100Gi,/mnt/disk3:disktype=SSD:capacity=100Gi,/mnt/disk4:disktype=SSD:capacity=100Gi
    celeborn.application.heartbeat.timeout=120s
    celeborn.master.ha.enabled=true
    celeborn.master.http.port=9098
    celeborn.metrics.enabled=true
    celeborn.push.stageEnd.timeout=120s
    celeborn.rpc.dispatcher.numThreads=4
    celeborn.rpc.io.clientThreads=64
    celeborn.rpc.io.numConnectionsPerPeer=2
    celeborn.rpc.io.serverThreads=64
    celeborn.shuffle.chunk.size=8m
    celeborn.worker.fetch.io.threads=32
    celeborn.worker.flusher.buffer.size=256K
    celeborn.worker.heartbeat.timeout=120s
    celeborn.worker.http.port=9096
    celeborn.worker.monitor.disk.enabled=false
    celeborn.worker.push.io.threads=32

  celeborn-env.sh: |
    CELEBORN_MASTER_JAVA_OPTS="-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
    CELEBORN_MASTER_MEMORY="2g"
    CELEBORN_NO_DAEMONIZE="1"
    CELEBORN_WORKER_JAVA_OPTS="-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
    CELEBORN_WORKER_MEMORY="2g"
    CELEBORN_WORKER_OFFHEAP_MEMORY="12g"
    TZ="Asia/Shanghai"

  log4j2.xml: |-
    <?xml version="1.0" encoding="UTF-8"?>
    <!--
    ~ Licensed to the Apache Software Foundation (ASF) under one or more
    ~ contributor license agreements.  See the NOTICE file distributed with
    ~ this work for additional information regarding copyright ownership.
    ~ The ASF licenses this file to You under the Apache License, Version 2.0
    ~ (the "License"); you may not use this file except in compliance with
    ~ the License.  You may obtain a copy of the License at
    ~
    ~     http://www.apache.org/licenses/LICENSE-2.0
    ~
    ~ Unless required by applicable law or agreed to in writing, software
    ~ distributed under the License is distributed on an "AS IS" BASIS,
    ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    ~ See the License for the specific language governing permissions and
    ~ limitations under the License.
    -->
    <!--
    ~ Extra logging related to initialization of Log4j.
    ~ Set to debug or trace if log4j initialization is failing.
    -->
    <Configuration status="INFO">
        <Appenders>
            <Console name="stdout" target="SYSTEM_OUT">
                <!--
                  ~ In the pattern layout configuration below, we specify an explicit `%ex` conversion
                  ~ pattern for logging Throwables. If this was omitted, then (by default) Log4J would
                  ~ implicitly add an `%xEx` conversion pattern which logs stacktraces with additional
                  ~ class packaging information. That extra information can sometimes add a substantial
                  ~ performance overhead, so we disable it in our default logging config.
                  -->
                <PatternLayout pattern="%d{yy/MM/dd HH:mm:ss,SSS} %p [%t] %c{1}: %m%n%ex"/>
            </Console>
            <RollingRandomAccessFile name="file" fileName="${env:CELEBORN_LOG_DIR}/celeborn.log"
                                     filePattern="${env:CELEBORN_LOG_DIR}/celeborn.log.%d-%i">
                <PatternLayout pattern="%d{yy/MM/dd HH:mm:ss,SSS} %p [%t] %c{1}: %m%n%ex"/>
                <Policies>
                    <SizeBasedTriggeringPolicy size="200 MB"/>
                </Policies>
                <DefaultRolloverStrategy max="7">
                    <Delete basePath="${env:CELEBORN_LOG_DIR}" maxDepth="1">
                        <IfFileName glob="celeborn.log*">
                            <IfAny>
                                <IfAccumulatedFileSize exceeds="1 GB" />
                                <IfAccumulatedFileCount exceeds="10" />
                            </IfAny>
                        </IfFileName>
                    </Delete>
                </DefaultRolloverStrategy>
            </RollingRandomAccessFile>
        </Appenders>

        <Loggers>
            <Root level="INFO">
                <AppenderRef ref="stdout"/>
                <AppenderRef ref="file"/>
            </Root>
            <Logger name="org.apache.hadoop.hdfs" level="WARN" additivity="false">
                <Appender-ref ref="stdout" level="WARN" />
                <Appender-ref ref="file" level="WARN"/>
            </Logger>
        </Loggers>
    </Configuration>

  metrics.properties: >-
    *.sink.prometheusServlet.class=org.apache.celeborn.common.metrics.sink.PrometheusServlet
---
# Source: celeborn/templates/master/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: celeborn-master-svc
  labels:
    helm.sh/chart: celeborn-0.1.0
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/version: "0.4.0"
    app.kubernetes.io/managed-by: Helm
spec:
  selector:
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/role: master
  ports:
  - port: 9097
    targetPort: 9097
    protocol: TCP
    name: celeborn-master
  type: ClusterIP
  clusterIP: None
---
# Source: celeborn/templates/worker/service.yaml
apiVersion: v1
kind: Service
metadata:
  name: celeborn-worker-svc
  labels:
    helm.sh/chart: celeborn-0.1.0
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/version: "0.4.0"
    app.kubernetes.io/managed-by: Helm
spec:
  selector:
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/role: worker
  type: ClusterIP
  clusterIP: None
---
# Source: celeborn/templates/master/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: celeborn-master
  labels:
    helm.sh/chart: celeborn-0.1.0
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/version: "0.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/role: master
spec:
  serviceName: celeborn-master-svc
  selector:
    matchLabels:
      app.kubernetes.io/name: celeborn
      app.kubernetes.io/instance: celeborn
      app.kubernetes.io/role: master
  template:
    metadata:
      labels:
        app.kubernetes.io/name: celeborn
        app.kubernetes.io/instance: celeborn
        app.kubernetes.io/role: master
        app.kubernetes.io/tag: "v0.4.0-incubating"
    spec:
      initContainers:
      - name: chown-celeborn-master-volume
        image: alpine:3.18
        imagePullPolicy: Never
        command:
        - chown
        - 10006:10006
        - /mnt/celeborn_ratis
        volumeMounts:
        - name: celeborn-master-vol-0
          mountPath: /mnt/celeborn_ratis
        securityContext:
          runAsUser: 0
      containers:
      - name: celeborn
        image: apache/celeborn:v0.4.0-incubating
        imagePullPolicy: Never
        command:
        - /usr/bin/tini
        - --
        - /bin/sh
        - -c
        - "until true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-master.sh"
        ports:
        - containerPort: 9097
        - containerPort: 9098
          name: metrics
          protocol: TCP
        env:
        - name: CELEBORN_MASTER_JAVA_OPTS
          value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
        - name: CELEBORN_MASTER_MEMORY
          value: "2g"
        - name: CELEBORN_NO_DAEMONIZE
          value: "1"
        - name: CELEBORN_WORKER_JAVA_OPTS
          value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
        - name: CELEBORN_WORKER_MEMORY
          value: "2g"
        - name: CELEBORN_WORKER_OFFHEAP_MEMORY
          value: "12g"
        - name: TZ
          value: "Asia/Shanghai"
        volumeMounts:
        - name: celeborn-volume
          mountPath: /opt/celeborn/conf
          readOnly: true
        - name: celeborn-master-vol-0
          mountPath: /mnt/celeborn_ratis
      volumes:
      - name: celeborn-volume
        configMap:
          name: celeborn-conf
      - name: celeborn-master-vol-0
        hostPath:
          path: /mnt/celeborn_ratis/master
          type: DirectoryOrCreate
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app.kubernetes.io/name
                operator: In
                values:
                - celeborn
              - key: app.kubernetes.io/role
                operator: In
                values:
                - master
            topologyKey: kubernetes.io/hostname
      dnsPolicy: ClusterFirst
      securityContext:
        fsGroup: 10006
        runAsGroup: 10006
        runAsUser: 10006
      terminationGracePeriodSeconds: 30
  replicas: 3
---
# Source: celeborn/templates/worker/statefulset.yaml
apiVersion: apps/v1
kind: StatefulSet
metadata:
  name: celeborn-worker
  labels:
    helm.sh/chart: celeborn-0.1.0
    app.kubernetes.io/name: celeborn
    app.kubernetes.io/instance: celeborn
    app.kubernetes.io/version: "0.4.0"
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/role: worker
spec:
  serviceName: celeborn-worker-svc
  selector:
    matchLabels:
      app.kubernetes.io/name: celeborn
      app.kubernetes.io/instance: celeborn
      app.kubernetes.io/role: worker
  template:
    metadata:
      labels:
        app.kubernetes.io/name: celeborn
        app.kubernetes.io/instance: celeborn
        app.kubernetes.io/role: worker
        app.kubernetes.io/tag: "v0.4.0-incubating"
    spec:
      initContainers:
      - name: chown-celeborn-worker-volume
        image: alpine:3.18
        imagePullPolicy: Never
        command:
        - chown
        - 10006:10006
        - /mnt/disk1
        - /mnt/disk2
        - /mnt/disk3
        - /mnt/disk4
        volumeMounts:
        - name: celeborn-worker-vol-0
          mountPath: /mnt/disk1
        - name: celeborn-worker-vol-1
          mountPath: /mnt/disk2
        - name: celeborn-worker-vol-2
          mountPath: /mnt/disk3
        - name: celeborn-worker-vol-3
          mountPath: /mnt/disk4
        securityContext:
          runAsUser: 0
      containers:
      - name: celeborn
        image: apache/celeborn:v0.4.0-incubating
        imagePullPolicy: Never
        command:
        - /usr/bin/tini
        - --
        - /bin/sh
        - -c
        - "until true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-worker.sh"
        ports:
        - containerPort: 9096
          name: metrics
          protocol: TCP
        env:
        - name: CELEBORN_MASTER_JAVA_OPTS
          value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
        - name: CELEBORN_MASTER_MEMORY
          value: "2g"
        - name: CELEBORN_NO_DAEMONIZE
          value: "1"
        - name: CELEBORN_WORKER_JAVA_OPTS
          value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
        - name: CELEBORN_WORKER_MEMORY
          value: "2g"
        - name: CELEBORN_WORKER_OFFHEAP_MEMORY
          value: "12g"
        - name: TZ
          value: "Asia/Shanghai"
        volumeMounts:
        - mountPath: /opt/celeborn/conf
          name: celeborn-volume
          readOnly: true
        - name: celeborn-worker-vol-0
          mountPath: /mnt/disk1
        - name: celeborn-worker-vol-1
          mountPath: /mnt/disk2
        - name: celeborn-worker-vol-2
          mountPath: /mnt/disk3
        - name: celeborn-worker-vol-3
          mountPath: /mnt/disk4
      volumes:
      - name: celeborn-volume
        configMap:
          name: celeborn-conf
      - name: celeborn-worker-vol-0
        hostPath:
          path: /mnt/disk1/worker
          type: DirectoryOrCreate
      - name: celeborn-worker-vol-1
        hostPath:
          path: /mnt/disk2/worker
          type: DirectoryOrCreate
      - name: celeborn-worker-vol-2
        hostPath:
          path: /mnt/disk3/worker
          type: DirectoryOrCreate
      - name: celeborn-worker-vol-3
        hostPath:
          path: /mnt/disk4/worker
          type: DirectoryOrCreate
      affinity:
        podAntiAffinity:
          requiredDuringSchedulingIgnoredDuringExecution:
          - labelSelector:
              matchExpressions:
              - key: app.kubernetes.io/name
                operator: In
                values:
                - celeborn
              - key: app.kubernetes.io/role
                operator: In
                values:
                - worker
            topologyKey: kubernetes.io/hostname
      dnsPolicy: ClusterFirst
      securityContext:
        fsGroup: 10006
        runAsGroup: 10006
        runAsUser: 10006
      terminationGracePeriodSeconds: 30
  replicas: 5

NOTES:
#
# Licensed to the Apache Software Foundation (ASF) under one or more
# contributor license agreements.  See the NOTICE file distributed with
# this work for additional information regarding copyright ownership.
# The ASF licenses this file to You under the Apache License, Version 2.0
# (the "License"); you may not use this file except in compliance with
# the License.  You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#

Celeborn
```

4. Wait a momemt and inspect the pod status

```shell
$ kubectl get pods -n celeborn
NAME                READY   STATUS    RESTARTS   AGE
celeborn-master-0   1/1     Running   0          44s
celeborn-master-1   1/1     Running   0          42s
celeborn-master-2   1/1     Running   0          40s
celeborn-worker-0   1/1     Running   0          44s
celeborn-worker-1   1/1     Running   0          42s
celeborn-worker-2   1/1     Running   0          40s
celeborn-worker-3   1/1     Running   0          38s
celeborn-worker-4   1/1     Running   0          35s
```

5. Inspect master logs

```shell
$ kubectl logs -n celeborn celeborn-master-0
...
24/05/10 16:58:46,907 WARN [main] JVMSource: Add gauge jvm.thread.deadlocks failed, the value type class java.util.Collections$EmptySet is not a number
24/05/10 16:58:46,952 WARN [main] ThreadPoolSource: Add gauge is_terminating failed, the value type class java.lang.Boolean is not a number
24/05/10 16:58:46,953 WARN [main] ThreadPoolSource: Add gauge is_terminated failed, the value type class java.lang.Boolean is not a number
24/05/10 16:58:46,953 WARN [main] ThreadPoolSource: Add gauge is_shutdown failed, the value type class java.lang.Boolean is not a number
24/05/10 16:58:46,954 INFO [main] Dispatcher: Dispatcher numThreads: 4
24/05/10 16:58:47,051 INFO [main] TransportClientFactory: mode NIO threads 64
24/05/10 16:58:47,151 WARN [main] ThreadPoolSource: Add gauge is_terminating failed, the value type class java.lang.Boolean is not a number
24/05/10 16:58:47,151 WARN [main] ThreadPoolSource: Add gauge is_terminated failed, the value type class java.lang.Boolean is not a number
24/05/10 16:58:47,151 WARN [main] ThreadPoolSource: Add gauge is_shutdown failed, the value type class java.lang.Boolean is not a number
24/05/10 16:58:47,162 INFO [main] NettyRpcEnvFactory: Starting RPC Server [Master] on celeborn-master-0.celeborn-master-svc.celeborn.svc.cluster.local:9097 with advisor endpoint celeborn-master-0.celeborn-master-svc.celeborn.svc.cluster.local:9097
24/05/10 16:58:47,306 INFO [main] Utils: Successfully started service 'Master' on port 9097.
24/05/10 16:58:47,577 INFO [main] RaftServer: raft.rpc.type = NETTY (custom)
24/05/10 16:58:47,682 INFO [main] NettyConfigKeys$Server: raft.netty.server.host = null (default)
24/05/10 16:58:47,684 INFO [main] NettyConfigKeys$Server: raft.netty.server.port = 9872 (custom)
24/05/10 16:58:47,686 INFO [main] DataStreamServerImpl: raft.datastream.type = DISABLED (default)
24/05/10 16:58:47,709 INFO [main] RaftServerConfigKeys: raft.server.threadpool.proxy.cached = true (default)
24/05/10 16:58:47,709 INFO [main] RaftServerConfigKeys: raft.server.threadpool.proxy.size = 0 (default)
24/05/10 16:58:47,710 INFO [main] RaftServerConfigKeys: raft.server.rpc.slowness.timeout = 120s (custom)
24/05/10 16:58:47,711 INFO [main] RaftServerConfigKeys: raft.server.leaderelection.leader.step-down.wait-time = 10s (default)
24/05/10 16:58:47,714 INFO [main] RaftServerConfigKeys: raft.server.storage.dir = [/mnt/celeborn_ratis] (custom)
...
```

6. Inspect the worker logs

```shell
$ kubectl logs -n celeborn celeborn-worker-0
...
24/05/10 16:59:19,697 INFO [main] MasterClient: connect to master celeborn-master-1.celeborn-master-svc.celeborn.svc.cluster.local:9097.
24/05/10 16:59:19,717 INFO [main] Worker: Register worker successfully.
24/05/10 16:59:19,723 INFO [main] PushDataHandler: diskReserveSize 5.0 GiB, diskReserveRatio null
24/05/10 16:59:19,724 INFO [main] PushDataHandler: diskReserveSize 5.0 GiB, diskReserveRatio null
24/05/10 16:59:19,725 INFO [main] Worker: Worker started.
24/05/10 16:59:27,568 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 8.0 MiB/12.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
24/05/10 16:59:37,576 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 8.0 MiB/12.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
24/05/10 16:59:47,585 INFO [worker-memory-manager-reporter] MemoryManager: Direct memory usage: 8.0 MiB/12.0 GiB, disk buffer size: 0.0 B, sort memory size: 0.0 B, read buffer size: 0.0 B
24/05/10 16:59:49,747 INFO [worker-forward-message-scheduler] StorageManager: Updated diskInfos:
DiskInfo(maxSlots: 0, committed shuffles 0, running applications 0, shuffleAllocations: Map(), mountPoint: /mnt/disk1, usableSpace: 32.2 GiB, avgFlushTime: 0 ns, avgFetchTime: 0 ns, activeSlots: 0, storageType: SSD) status: HEALTHY dirs /mnt/disk1/celeborn-worker/shuffle_data
DiskInfo(maxSlots: 0, committed shuffles 0, running applications 0, shuffleAllocations: Map(), mountPoint: /mnt/disk3, usableSpace: 32.2 GiB, avgFlushTime: 0 ns, avgFetchTime: 0 ns, activeSlots: 0, storageType: SSD) status: HEALTHY dirs /mnt/disk3/celeborn-worker/shuffle_data
DiskInfo(maxSlots: 0, committed shuffles 0, running applications 0, shuffleAllocations: Map(), mountPoint: /mnt/disk2, usableSpace: 32.2 GiB, avgFlushTime: 0 ns, avgFetchTime: 0 ns, activeSlots: 0, storageType: SSD) status: HEALTHY dirs /mnt/disk2/celeborn-worker/shuffle_data
DiskInfo(maxSlots: 0, committed shuffles 0, running applications 0, shuffleAllocations: Map(), mountPoint: /mnt/disk4, usableSpace: 32.2 GiB, avgFlushTime: 0 ns, avgFetchTime: 0 ns, activeSlots: 0, storageType: SSD) status: HEALTHY dirs /mnt/disk4/celeborn-worker/shuffle_data
24/05/10 16:59:49,862 INFO [worker-expired-shuffle-cleaner] ChunkStreamManager: Clean up expired shuffle keys
...
```

RexXiong commented:
Thanks @ChenYi015 for this pr, Good job. cc @zwangsheng

ChenYi015 commented on charts/celeborn/tests/master/podmonitor_test.yaml:
```diff
@@ -0,0 +1,72 @@
+#
```
The files located in `charts/celeborn/templates` are in fact Go template files, hence their license is commented using `{{/* */}}`. The remaining YAML files are standard YAML files, and therefore, their license is commented with `#`.

RexXiong commented on charts/celeborn/README.md:
```diff
@@ -17,19 +17,20 @@ specific language governing permissions and limitations
-[Apache Celeborn](https://celeborn.apache.org) is an intermediate data service for Big Data compute engines (i.e. ETL, OLAP and Streaming engines) to boost performance, stability, and flexibility. Intermediate data typically include shuffle and spilled data.
+![Version: 0.1.0](https://img.shields.io/badge/Version-0.1.0-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 0.4.0](https://img.shields.io/badge/AppVersion-0.4.0-informational?style=flat-square)
 
+Celeborn is dedicated to improving the efficiency and elasticity of different map-reduce engines. Celeborn provides an elastic and high efficient management service for shuffle data.
```
Better keep this, and the content here should be consistent with the content on the Celeborn website see [Apache Celeborn](https://celeborn.apache.org/)

ChenYi015 replied:
> Better keep this, and the content here should be consistent with the content on the Celeborn website see [Apache Celeborn](https://celeborn.apache.org/)

Updated.

RexXiong commented on charts/celeborn/README.md:
```diff
@@ -40,17 +41,57 @@ When you want to test the template rendering, but not actually install anything.
+| service.type | string | `"ClusterIP"` | Specifies service type |
+| tolerations | list | `[]` | Pod tolerations |
+| volumes | object | `{}` | Specifies Celeborn volumes. Current Celeborn support followings volume types are `emptyDir` and `hostPath`. <br> Note that `hostPath` only works in hostPath type using to set `volumes hostPath path`. Celeborn Master will pick first volumes for store raft log. <br> `diskType` only works in Celeborn Worker with hostPath type to manifest local disk type |
+| workerReplicas | int | `5` | Specifies the number of Celeborn worker replicas to deploy, should less than node number |
```
ditto.

RexXiong commented on charts/celeborn/README.md:
```diff
@@ -40,17 +41,57 @@ When you want to test the template rendering, but not actually install anything.
+| image.repository | string | `"aliyunemr/remote-shuffle-service"` | Image repository |
+| image.tag | string | `"0.1.1-6badd20"` | Image tag |
+| imagePullSecrets | list | `[]` | Image pull secrets for private image registry |
+| masterReplicas | int | `3` | Specifies the number of Celeborn master replicas to deploy, master replicas should not less than 3 |
```
Why the key is still masterReplicas?

ChenYi015 replied:
> Why the key is still masterReplicas?

Sorry, I forget to run [helm-docs](https://github.com/norwoodj/helm-docs) go generate the latest documentation. And a pre-commit git hook can be configured to run this command automatically before commit.

pan3793 replied:
what if the developer does not have [helm-docs](https://github.com/norwoodj/helm-docs) installed locally?

RexXiong reviewed (commented):
Nice pr, I have some comments left. In addition, can you compare the actual YAML files before and after the modifications to see what differences there are?

ChenYi015 commented:
> Nice pr, I have some comments left. In addition, can you compare the actual YAML files before and after the modifications to see what differences there are?

```shell
$ git checkout main
Switched to branch 'main'
$ helm template celeborn charts/celeborn >original.yaml
(base) 
$ git checkout helm-unittest
Switched to branch 'helm-unittest'
Your branch is up to date with 'origin/helm-unittest'.
(base) 
$ helm template celeborn charts/celeborn >current.yaml
(base) 
$ diff original.yaml current.yaml
3,19d2
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< 
132,149c115
< # Source: celeborn/templates/master-service.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< 
---
> # Source: celeborn/templates/master/service.yaml
161,167d126
<   type: ClusterIP
<   ports:
<     - port: 9097
<       targetPort: 9097
<       protocol: TCP
<       name: celeborn-master
<   clusterIP: None
171a131,137
>   ports:
>   - port: 9097
>     targetPort: 9097
>     protocol: TCP
>     name: celeborn-master
>   type: ClusterIP
>   clusterIP: None
173,190c139
< # Source: celeborn/templates/worker-service.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< 
---
> # Source: celeborn/templates/worker/service.yaml
202,203d150
<   type: ClusterIP
<   clusterIP: None
207a155,156
>   type: ClusterIP
>   clusterIP: None
209,226c158
< # Source: celeborn/templates/master-statefulset.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< 
---
> # Source: celeborn/templates/master/statefulset.yaml
238a171
>   serviceName: celeborn-master-svc
244,245d176
<   serviceName: celeborn-master-svc
<   replicas: 3
253a185,241
>       initContainers:
>       - name: chown-celeborn-master-volume
>         image: alpine:3.18
>         imagePullPolicy: Always
>         command:
>         - chown
>         - 10006:10006
>         - /mnt/celeborn_ratis
>         volumeMounts:
>         - name: celeborn-master-vol-0
>           mountPath: /mnt/celeborn_ratis
>         securityContext:
>           runAsUser: 0
>       containers:
>       - name: celeborn
>         image: aliyunemr/remote-shuffle-service:0.1.1-6badd20
>         imagePullPolicy: Always
>         command:
>         - /usr/bin/tini
>         - --
>         - /bin/sh
>         - -c
>         - "until true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-master.sh"
>         ports:
>         - containerPort: 9097
>         - containerPort: 9098
>           name: metrics
>           protocol: TCP
>         env:
>         - name: CELEBORN_MASTER_JAVA_OPTS
>           value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
>         - name: CELEBORN_MASTER_MEMORY
>           value: "2g"
>         - name: CELEBORN_NO_DAEMONIZE
>           value: "1"
>         - name: CELEBORN_WORKER_JAVA_OPTS
>           value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
>         - name: CELEBORN_WORKER_MEMORY
>           value: "2g"
>         - name: CELEBORN_WORKER_OFFHEAP_MEMORY
>           value: "12g"
>         - name: TZ
>           value: "Asia/Shanghai"
>         volumeMounts:
>         - name: celeborn-volume
>           mountPath: /opt/celeborn/conf
>           readOnly: true
>         - name: celeborn-master-vol-0
>           mountPath: /mnt/celeborn_ratis
>       volumes:
>       - name: celeborn-volume
>         configMap:
>           name: celeborn-conf
>       - name: celeborn-master-vol-0
>         hostPath:
>           path: /mnt/celeborn_ratis/master
>           type: DirectoryOrCreate
267a256
>       dnsPolicy: ClusterFirst
272,326d260
<       hostNetwork: false
<       dnsPolicy: ClusterFirst
<       initContainers:
<       - name: chown-celeborn-master-volume
<         image: alpine:3.18
<         imagePullPolicy: Always
<         securityContext:
<           runAsUser: 0
<         command:
<         - chown
<         - 10006:10006
<         - /mnt/celeborn_ratis
<         resources:
<             null
<         volumeMounts:
<           - name: celeborn-master-vol-0
<             mountPath: /mnt/celeborn_ratis
<       containers:
<       - name: celeborn
<         image: "aliyunemr/remote-shuffle-service:0.1.1-6badd20"
<         imagePullPolicy: Always
<         command:
<           - "/usr/bin/tini"
<           - "--"
<           - "/bin/sh"
<           - '-c'
<           - "until nslookup celeborn-master-0.celeborn-master-svc.default.svc.cluster.local && nslookup celeborn-master-1.celeborn-master-svc.default.svc.cluster.local && nslookup celeborn-master-2.celeborn-master-svc.default.svc.cluster.local && true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-master.sh"
<         resources:
<             null
<         ports:
<           - containerPort: 9097
<           - containerPort: 9098
<             name: metrics
<             protocol: TCP
<         volumeMounts:
<           - mountPath: /opt/celeborn/conf
<             name: celeborn-volume
<             readOnly: true
<           - name: celeborn-master-vol-0
<             mountPath: /mnt/celeborn_ratis
<         env:
<           - name: CELEBORN_MASTER_JAVA_OPTS
<             value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
<           - name: CELEBORN_MASTER_MEMORY
<             value: "2g"
<           - name: CELEBORN_NO_DAEMONIZE
<             value: "1"
<           - name: CELEBORN_WORKER_JAVA_OPTS
<             value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
<           - name: CELEBORN_WORKER_MEMORY
<             value: "2g"
<           - name: CELEBORN_WORKER_OFFHEAP_MEMORY
<             value: "12g"
<           - name: TZ
<             value: "Asia/Shanghai"
328,335c262
<       volumes:
<         - configMap:
<             name: celeborn-conf
<           name: celeborn-volume
<         - name: celeborn-master-vol-0
<           hostPath:
<             path: /mnt/celeborn_ratis/master
<             type: DirectoryOrCreate
---
>   replicas: 3
337,354c264
< # Source: celeborn/templates/worker-statefulset.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< 
---
> # Source: celeborn/templates/worker/statefulset.yaml
366a277
>   serviceName: celeborn-worker-svc
372,373d282
<   serviceName: celeborn-worker
<   replicas: 5
382,401d290
<       affinity:
<         podAntiAffinity:
<           requiredDuringSchedulingIgnoredDuringExecution:
<           - labelSelector:
<               matchExpressions:
<               - key: app.kubernetes.io/name
<                 operator: In
<                 values:
<                 - celeborn
<               - key: app.kubernetes.io/role
<                 operator: In
<                 values:
<                 - worker
<             topologyKey: kubernetes.io/hostname
<       securityContext:
<         fsGroup: 10006
<         runAsGroup: 10006
<         runAsUser: 10006
<       hostNetwork: false
<       dnsPolicy: ClusterFirst
406,407d294
<         securityContext:
<           runAsUser: 0
415,416d301
<         resources:
<             null
425a311,312
>         securityContext:
>           runAsUser: 0
428c315
<         image: "aliyunemr/remote-shuffle-service:0.1.1-6badd20"
---
>         image: aliyunemr/remote-shuffle-service:0.1.1-6badd20
431,437c318,322
<           - "/usr/bin/tini"
<           - "--"
<           - "/bin/sh"
<           - '-c'
<           - "until nslookup celeborn-master-0.celeborn-master-svc.default.svc.cluster.local && nslookup celeborn-master-1.celeborn-master-svc.default.svc.cluster.local && nslookup celeborn-master-2.celeborn-master-svc.default.svc.cluster.local && true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-worker.sh"
<         resources:
<             null
---
>         - /usr/bin/tini
>         - --
>         - /bin/sh
>         - -c
>         - "until true; do echo waiting for master; sleep 2; done && exec /opt/celeborn/sbin/start-worker.sh"
439,453c324,326
<           - containerPort: 9096
<             name: metrics
<             protocol: TCP
<         volumeMounts:
<           - mountPath: /opt/celeborn/conf
<             name: celeborn-volume
<             readOnly: true
<           - name: celeborn-worker-vol-0
<             mountPath: /mnt/disk1
<           - name: celeborn-worker-vol-1
<             mountPath: /mnt/disk2
<           - name: celeborn-worker-vol-2
<             mountPath: /mnt/disk3
<           - name: celeborn-worker-vol-3
<             mountPath: /mnt/disk4
---
>         - containerPort: 9096
>           name: metrics
>           protocol: TCP
455,472c328,343
<           - name: CELEBORN_MASTER_JAVA_OPTS
<             value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
<           - name: CELEBORN_MASTER_MEMORY
<             value: "2g"
<           - name: CELEBORN_NO_DAEMONIZE
<             value: "1"
<           - name: CELEBORN_WORKER_JAVA_OPTS
<             value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
<           - name: CELEBORN_WORKER_MEMORY
<             value: "2g"
<           - name: CELEBORN_WORKER_OFFHEAP_MEMORY
<             value: "12g"
<           - name: TZ
<             value: "Asia/Shanghai"
<       terminationGracePeriodSeconds: 30
<       volumes:
<         - configMap:
<             name: celeborn-conf
---
>         - name: CELEBORN_MASTER_JAVA_OPTS
>           value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-master.out -Dio.netty.leakDetectionLevel=advanced"
>         - name: CELEBORN_MASTER_MEMORY
>           value: "2g"
>         - name: CELEBORN_NO_DAEMONIZE
>           value: "1"
>         - name: CELEBORN_WORKER_JAVA_OPTS
>           value: "-XX:-PrintGC -XX:+PrintGCDetails -XX:+PrintGCTimeStamps -XX:+PrintGCDateStamps -Xloggc:gc-worker.out -Dio.netty.leakDetectionLevel=advanced"
>         - name: CELEBORN_WORKER_MEMORY
>           value: "2g"
>         - name: CELEBORN_WORKER_OFFHEAP_MEMORY
>           value: "12g"
>         - name: TZ
>           value: "Asia/Shanghai"
>         volumeMounts:
>         - mountPath: /opt/celeborn/conf
473a345
>           readOnly: true
475,477c347
<           hostPath:
<             path: /mnt/disk1/worker
<             type: DirectoryOrCreate
---
>           mountPath: /mnt/disk1
479,481c349
<           hostPath:
<             path: /mnt/disk2/worker
<             type: DirectoryOrCreate
---
>           mountPath: /mnt/disk2
483,485c351
<           hostPath:
<             path: /mnt/disk3/worker
<             type: DirectoryOrCreate
---
>           mountPath: /mnt/disk3
487,543c353,394
<           hostPath:
<             path: /mnt/disk4/worker
<             type: DirectoryOrCreate
< ---
< # Source: celeborn/templates/master-priorityclass.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< ---
< # Source: celeborn/templates/prometheus-podmonitor.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
< ---
< # Source: celeborn/templates/worker-priorityclass.yaml
< #
< # Licensed to the Apache Software Foundation (ASF) under one or more
< # contributor license agreements.  See the NOTICE file distributed with
< # this work for additional information regarding copyright ownership.
< # The ASF licenses this file to You under the Apache License, Version 2.0
< # (the "License"); you may not use this file except in compliance with
< # the License.  You may obtain a copy of the License at
< #
< #    http://www.apache.org/licenses/LICENSE-2.0
< #
< # Unless required by applicable law or agreed to in writing, software
< # distributed under the License is distributed on an "AS IS" BASIS,
< # WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
< # See the License for the specific language governing permissions and
< # limitations under the License.
< #
---
>           mountPath: /mnt/disk4
>       volumes:
>       - name: celeborn-volume
>         configMap:
>           name: celeborn-conf
>       - name: celeborn-worker-vol-0
>         hostPath:
>           path: /mnt/disk1/worker
>           type: DirectoryOrCreate
>       - name: celeborn-worker-vol-1
>         hostPath:
>           path: /mnt/disk2/worker
>           type: DirectoryOrCreate
>       - name: celeborn-worker-vol-2
>         hostPath:
>           path: /mnt/disk3/worker
>           type: DirectoryOrCreate
>       - name: celeborn-worker-vol-3
>         hostPath:
>           path: /mnt/disk4/worker
>           type: DirectoryOrCreate
>       affinity:
>         podAntiAffinity:
>           requiredDuringSchedulingIgnoredDuringExecution:
>           - labelSelector:
>               matchExpressions:
>               - key: app.kubernetes.io/name
>                 operator: In
>                 values:
>                 - celeborn
>               - key: app.kubernetes.io/role
>                 operator: In
>                 values:
>                 - worker
>             topologyKey: kubernetes.io/hostname
>       dnsPolicy: ClusterFirst
>       securityContext:
>         fsGroup: 10006
>         runAsGroup: 10006
>         runAsUser: 10006
>       terminationGracePeriodSeconds: 30
>   replicas: 5
```

ChenYi015 commented on charts/celeborn/templates/worker/statefulset.yaml:
```diff
@@ -0,0 +1,149 @@
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- if or .Values.priorityClass.worker.name .Values.priorityClass.worker.create }}
+      priorityClassName: {{ .Values.priorityClass.worker.name | default (include "celeborn.masterPriorityClassName" .) }}
```
> "celeborn.workerPriorityClassName"

Fixed.

ChenYi015 commented on charts/celeborn/templates/master/statefulset.yaml:
```diff
@@ -0,0 +1,146 @@
+        {{- toYaml . | nindent 8 }}
+      {{- end }}
+      {{- if or .Values.priorityClass.master.name .Values.priorityClass.master.create }}
+      priorityClassName: {{ .Values.priorityClass.master.name | default (include "celeborn.workerPriorityClassName" .) }}
```
> include "celeborn.masterPriorityClassName"

Fixed

RexXiong reviewed (approved):
LGTM, thanks!

zwangsheng reviewed (approved):
Great work! LGTM

SteNicholas commented on charts/celeborn/README.md:
```diff
@@ -17,19 +17,22 @@ specific language governing permissions and limitations
 -->
 
-# Helm Chart for Apache Celeborn
+# celeborn
```
Revert to `Helm Chart for Apache Celeborn`.

SteNicholas commented on charts/celeborn/README.md:
```diff
@@ -17,19 +17,22 @@ specific language governing permissions and limitations
-[Apache Celeborn](https://celeborn.apache.org) is an intermediate data service for Big Data compute engines (i.e. ETL, OLAP and Streaming engines) to boost performance, stability, and flexibility. Intermediate data typically include shuffle and spilled data.
+![Version: 0.1.0](https://img.shields.io/badge/Version-0.1.0-informational?style=flat-square) ![Type: application](https://img.shields.io/badge/Type-application-informational?style=flat-square) ![AppVersion: 0.4.0](https://img.shields.io/badge/AppVersion-0.4.0-informational?style=flat-square)
 
+Celeborn is an intermediate data service for Big Data compute engines (i.e. ETL, OLAP and Streaming engines) to boost performance, stability, and flexibility. Intermediate data typically include shuffle and spilled data.
```
Line 24-26 revert to `[Apache Celeborn](https://celeborn.apache.org) is an intermediate data service for Big Data compute engines (i.e. ETL, OLAP and Streaming engines) to boost performance, stability, and flexibility. Intermediate data typically include shuffle and spilled data.`.

SteNicholas commented on charts/celeborn/README.md:
```diff
@@ -40,17 +43,57 @@ When you want to test the template rendering, but not actually install anything.
+| securityContext.fsGroup | int | `10006` | Specifies the group ID to use when modifying ownership and permissions of the mounted volumes |
+| securityContext.runAsGroup | int | `10006` | Specifies the group ID to run the entrypoint of the container process |
+| securityContext.runAsUser | int | `10006` | Specifies the user ID to run the entrypoint of the container process |
+| service.port | int | `9097` | Specifies service port |
```
Could this port distingish the port of the Master and Worker service?

ChenYi015 replied:
> Could this port distingish the port of the Master and Worker service?

Actually, `service.port` is used only by the master service, the worker service does not expose a service port.

SteNicholas replied:
@ChenYi015, could the worker service expose a service port?

ChenYi015 replied:
@SteNicholas Just wondering, why do we need to expose a port for the worker service?

SteNicholas replied:
@ChenYi015, could the users deploy worker service in other k8s cluster?

ChenYi015 replied:
@SteNicholas Currently not, we need to modify the helm chart so that we can deploy celeborn workers in another k8s cluster.

SteNicholas replied:
@ChenYi015, would you like to support this in another pull request?

ChenYi015 replied:
@SteNicholas If that's you need, I will try to make it happen.

SteNicholas commented on charts/celeborn/README.md:
```diff
@@ -40,17 +43,57 @@ When you want to test the template rendering, but not actually install anything.
+| podMonitor.podMetricsEndpoint | object | `{"interval":"5s","portName":"metrics","scheme":"http"}` | Specifies PodMonitor endpoint |
+| priorityClass.master.create | bool | `false` | Specifies whether a new priority class for Celeborn master pods should be created |
+| priorityClass.master.name | string | `""` | Specifies the name of priority class for Celeborn master pods to be used (created if `create: true`) |
+| priorityClass.master.value | int | `1000000000` | Specifies the integer value of this priority class, default is half of system-cluster-critical |
```
Could the description mention which values the priority class has?

ChenYi015 replied:
> Could the description mention which values the priority class has?

Added `empty means "${Release.Name}-master-priority-class"` in description.

SteNicholas commented on charts/celeborn/README.md:
```diff
@@ -40,17 +43,57 @@ When you want to test the template rendering, but not actually install anything.
+| nameOverride | string | `""` | String to override the default generated name |
+| nodeSelector | object | `{}` | Pod node selector |
+| podAnnotations | object | `{}` | Pod annotations |
+| podMonitor.enable | bool | `true` | Specifies whether a PodMonitor should be created |
```
```suggestion
| podMonitor.enable | bool | `true` | Specifies whether to enable creating a PodMonitor |
```

SteNicholas commented on charts/celeborn/templates/master/podmonitor.yaml:
```diff
@@ -0,0 +1,41 @@
+  - interval: {{ .Values.podMonitor.podMetricsEndpoint.interval }}
+    port: {{ .Values.podMonitor.podMetricsEndpoint.portName | quote }}
+    scheme: {{ .Values.podMonitor.podMetricsEndpoint.scheme }}
+    path: /metrics/prometheus
```
```suggestion
    path: {{ get .Values.celeborn "celeborn.metrics.prometheus.path" }}
```

ChenYi015 replied:
Added `{{ get .Values.celeborn "celeborn.metrics.prometheus.path" | default "/metrics/prometheus" }}`

pan3793 commented:
This PR involves refactoring, user-facing variable name changes, bug fixes, and testing hardening. Can you split it into some dedicated patches?

pan3793 commented:
> - Format yaml template files with kubernetes style.

Can you leave a docs/guide link?

ChenYi015 commented:
> This PR involves refactoring, user-facing variable name changes, bug fixes, and testing hardening. Can you split it into some dedicated patches?

@pan3793 OK, I will break it down into serveral smaller PRs.

ChenYi015 commented:
> > * Format yaml template files with kubernetes style.
> 
> Can you leave a docs/guide link?

@pan3793 There isn't any offical docs about Kubernetes format style now. But as a convetion, the sequence style in the k8s style yaml has no indent. For example, [Deployments | Kubernetes](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/#creating-a-deployment).
And each time you use [kubectl](https://kubernetes.io/docs/reference/kubectl/) to get manifest as yaml format, there is no indent.

pan3793 commented:
it's good if only about indention, but I would like to avoid such user-facing breaking changes unless you have good reasons.

```
- Rename `imagePullSecrets` value to `image.pullSecrets`
- Rename `masterReplicas` to `replicaCount.master`
- Rename `workerReplicas` to `replicaCount.worker`
```

ChenYi015 commented:
> it's good if only about indention, but I would like to avoid such user-facing breaking changes unless you have good reasons.
> 
> ```
> - Rename `imagePullSecrets` value to `image.pullSecrets`
> - Rename `masterReplicas` to `replicaCount.master`
> - Rename `workerReplicas` to `replicaCount.worker`
> ```

Maybe we can mark these fields as deprecated and provide back-forward compatibility in helm charts, how do you think about it?

pan3793 commented:
> Maybe we can mark these fields as deprecated and provide backward compatibility in helm charts, how do you think about it?

@ChenYi015 approach is good but you should provide **reasons** for renaming.

ChenYi015 commented:
@pan3793 For the number of replicas, `replicaCount` is a more common name, when you create a new helm chart, you will see `replicaCount` is used. Just like we have `volumes.[master|worker]`, `priorityClass.[master|worker]` and `affinity.[master|worker]`, `replicaCount.[master|worker]` will make the values much clearer.

ChenYi015 commented:
This PR had breaked into several smaller ones and will be cloesd.
